### DISCUSSION TITLE
Using BluetoothSerial.h with arduino-audio-tools

### DISCUSSION THREAD

**Discussion by rbagdazian** [pschatzmann/arduino-audio-tools#1713] [Q&A] [answered]
I have had great success in using your ESP32-A2DP and arduino-audio-tools libraries to implement a bluetooth based sound system with some additional internal features. I would like to be able to create an android/IOS app that communicates via BluetoothSerial for controlling some of the functionality in my system. Can BluetoothSerial protocol be used simultaneously when the ESP32-A2DP stack is running to support audio transport from a smartphone? Any thoughts on this are greatly appreciated!

**Comment by pschatzmann** [comment, marked as answer]
Just give it a try: You might need to call set_spp_active(true) first

**Reply by rbagdazian** [reply]
I'll give it a go, and report back what I find out. Thank you!

**Reply by pschatzmann** [reply]
A small side note: A2DP and BLE are wroking properly together and this has been tested.
So this might be another alternative...

**Reply by rbagdazian** [reply]
Are there any examples? I haven't worked with ble before. Thank you!

**Reply by rbagdazian** [reply]
No need for ble examples. I got spp working while simultaneously streaming audio to my board. I did call set_spp_active(true) 
Then made a call to start an instance of BluetoothSerial.  After that it worked fine with a simple loop back sequence in the Arduino loop function.

**Reply by rbagdazian** [reply]
I have discovered something interesting and I thought I would get your opinion on it. As I mentioned earlier, I have been able to get spp working just fine when using ESP32-A2DP from your library together with the arduino-audio-tools. In this case I did what you mentioned and call set_spp_active(true) along with starting the BluetoothSerial object. My project has another operating mode in which I accept audio input from an external I2S adc (a PCM1808) and route it to the same I2S port for output to an external dac (pcm5102). This is also working just fine using the arduino tools in your library. I wanted to also be able to connect to the board when in this mode using spp, however, when I enabled the BluetoothSerial object the system crashes and reboots with a panic abort of some sort. In this case I have not instantiated a BluetoothA2DPSink or anything related to it. Just the I2s audio configuration for master and RXTX mode, and the BluetoothSerial object by itself. Also I found that if I didn't initialize the I2S interface for audio processing I was able to communicate with the board via spp. Is there something when using i2s->i2s on the ESP32 that prevents using BluetoothSerial that you know of?
Thanks very much!

**Reply by pschatzmann** [reply]
Why don't you just analyse the error message and backtrace ? 
I guess you are just running out of resources (e.g. RAM)

**Reply by rbagdazian** [reply]
Yes, it may be a resources issue. The crash dump shows that the failure is occurring somewhere deep in the bluetooth stack. I thought it was odd, since when using ESP32-A2DP together with the arduino-audio-tools things are working fine, but when not using the A2DP part of the bluetooth stack, and instead using both sides of the I2S channel in master mode, the problem arises. I would have thought that the a2dp bt stack would use a lot more memory resources than the i2s drivers would. I'll spend some time digging into the esp bluetooth code where the failure is occurring to see what's happening. Thanks!

**Reply by pschatzmann** [reply]
You might also try if [using PSRAM helps](https://www.pschatzmann.ch/home/2022/05/30/esp32-and-psram/)

**Reply by rbagdazian** [reply]
I'll give that a try. I've used the heap_caps_malloc function elsewhere for some of my custom code to avoid these kinds of memory issues. Thank you!

**Reply by rbagdazian** [reply]
That fixed the problem! I had to use a fairly small value for the limit (1024) but everything is working as it should be now.
Thank you again!
